sdk/eventhubs/microsoft-azure-eventhubs/src/main/java/com/microsoft/azure/eventhubs/impl/MessagingFactory.java

```diff
@@ -856,29 +856,34 @@ public void run() {
         }
 
         private void scheduleCompletePendingTasks() {
-            this.executor.schedule(new Runnable() {
-                @Override
-                public void run() {
-                    if (TRACE_LOGGER.isWarnEnabled()) {
-                        TRACE_LOGGER.warn(String.format(Locale.US, "messagingFactory[%s], hostName[%s], message[%s]",
+            try {
+                this.executor.schedule(new Runnable() {
+                    @Override
+                    public void run() {
+                        if (TRACE_LOGGER.isWarnEnabled()) {
+                            TRACE_LOGGER.warn(String.format(Locale.US, "messagingFactory[%s], hostName[%s], message[%s]",
                                 getClientId(), getHostName(),
                                 "Processing all pending tasks and closing old reactor."));
-                    }
+                        }
 
-                    try {
-                        rctr.stop();
-                        rctr.process();
-                    } catch (HandlerException e) {
-                        if (TRACE_LOGGER.isWarnEnabled()) {
-                            TRACE_LOGGER.warn(String.format(Locale.US, "messagingFactory[%s], hostName[%s], error[%s]",
+                        try {
+                            rctr.stop();
+                            rctr.process();
+                        } catch (HandlerException e) {
+                            if (TRACE_LOGGER.isWarnEnabled()) {
+                                TRACE_LOGGER.warn(String.format(Locale.US, "messagingFactory[%s], hostName[%s], error[%s]",
                                     getClientId(), getHostName(), ExceptionUtil.toStackTraceString(e,
-                                            "scheduleCompletePendingTasks - exception occurred while processing events.")));
+                                        "scheduleCompletePendingTasks - exception occurred while processing events.")));
+                            }
+                        } finally {
+                            rctr.free();
                         }
-                    } finally {
-                        rctr.free();
                     }
-                }
-            }, MessagingFactory.this.getOperationTimeout().getSeconds(), TimeUnit.SECONDS);
+                }, MessagingFactory.this.getOperationTimeout().getSeconds(), TimeUnit.SECONDS);
+            } catch (Throwable t) {
+                TRACE_LOGGER.error("Failed to schedule cleanup task; freeing reactor synchronously", t);
+                rctr.free();
```

Comment:
the try block does `stop()` and `process()` on rctr so do we need to do that too?

```
                            rctr.stop();
                            rctr.process();
```

Reply:
i didnt go through the reactor code fully, but we use the executor to thread to run `rctr.stop() and  rctr.process()`. the reactor.process is an infinite loop which keep processing the events. so i don't think we need to call these methods in the catch block.

Reply:
got it